src/server/scripts/Northrend/Naxxramas/boss_noth.cpp

```diff
@@ -151,72 +238,119 @@ class boss_noth : public CreatureScript
+                    }
                     case EVENT_WARRIOR:
                         Talk(SAY_SUMMON);
-                        SummonUndead(NPC_WARRIOR, RAID_MODE(2, 3));
-                        events.ScheduleEvent(EVENT_WARRIOR, 30000);
-                        return;
+                        Talk(EMOTE_SUMMON);
+                        
+                        CastSummon(RAID_MODE(2, 3), 0, 0);
+
+                        events.ScheduleEvent(EVENT_WARRIOR, 40 * IN_MILLISECONDS, 0, PHASE_GROUND);
+                        break;
                     case EVENT_BLINK:
                         DoCastAOE(SPELL_CRIPPLE, true);
-                        DoCastAOE(SPELL_BLINK);
+                        DoCast(SPELL_BLINK);
                         DoResetThreat();
-                        events.ScheduleEvent(EVENT_BLINK, 40000);
-                        return;
+                        justBlinked = true;
+
+                        events.ScheduleEvent(EVENT_BLINK, 40000, 0, PHASE_GROUND);
+                        break;
                     case EVENT_BALCONY:
+                        events.SetPhase(PHASE_BALCONY);
                         me->SetReactState(REACT_PASSIVE);
-                        me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE);
+                        me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_PC | UNIT_FLAG_NOT_SELECTABLE);
                         me->AttackStop();
+                        me->StopMoving();
                         me->RemoveAllAuras();
-                        me->NearTeleportTo(Teleport.GetPositionX(), Teleport.GetPositionY(), Teleport.GetPositionZ(), Teleport.GetOrientation());
-                        events.Reset();
-                        events.ScheduleEvent(EVENT_WAVE, urand(2000, 5000));
-                        waveCount = 0;
-                        return;
+
+                        events.ScheduleEvent(EVENT_BALCONY_TELEPORT, 3 * IN_MILLISECONDS, 0, PHASE_BALCONY);
+                        events.ScheduleEvent(EVENT_WAVE, urand(5 * IN_MILLISECONDS, 8 * IN_MILLISECONDS), 0, PHASE_BALCONY);
+
+                        uint8 secondsBalcony;
+                        switch (balconyCount)
+                        {
+                            case 0:
+                                secondsBalcony = 70;
+                                break;
+                            case 1:
+                                secondsBalcony = 97;
+                                break;
+                            case 2:
+                            default:
+                                secondsBalcony = 120;
+                                break;
+                        }
+                        events.ScheduleEvent(EVENT_GROUND, secondsBalcony * IN_MILLISECONDS, 0, PHASE_BALCONY);
+                        break;
+                    case EVENT_BALCONY_TELEPORT:
+                        Talk(EMOTE_TELEPORT_1);
+                        DoCast(SPELL_TELEPORT);
```

Comment:
https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/game/AI/CoreAI/UnitAI.cpp#L93

https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/game/AI/CoreAI/UnitAI.cpp#L143

DoCast(spellid) is used by basic AI implementations (iirc SAI) so should be avoided